src/routes/guides/testing-your-app.mdx

```diff
@@ -0,0 +1,404 @@
+
+test("toggles fullscreen", () => {
+  const targetElement = document.createElement("div")
+  const fs = vi.spyOn(targetElement, "fullscreen")
+  const [setArg, container] = renderDirective(createFullScreen, false)
+  setArg(true)
+  expect(fs).toHaveBeenCalled()
+})
+```
+  </div>
+  <div id="fullscreen.ts">
+```ts frame="none"
+import { Accessor } from "solid-js"
+
+export const fullscreen = (ref: HTMLElement, active: Accessor<boolean>) =>
+  createEffect(() => {
+    const isActive = document.fullscreenElement === ref
+    if (active() && !isActive) {
+      ref.requestFullScreen().catch(() => {})
+    } else if (!active() && isActive) {
+      document.exitFullScreen()
+    }
+  })
+```
+  </div>
+</TabsCodeBlocks>
+
+## Primitive testing
+
+If the reference to the element is not needed, parts of state and logic can be put into reusable hooks or primitives. Since those do not require any elements, using `render` to test them would require a component that otherwise has no use. To circumvent that, there is a `renderHook` utility that simulates such a component without actually rendering anything.
+
+```ts frame="none"
+const renderResult = renderHook(hook, {
+  initialProps, // an array with arguments being supplied to the hook
+  wrapper, // same as the wrapper optionss for `render`
+})
+const {
+  result, // return value of the hook (mutable, destructuring fixes it)
+  cleanup, // manually remove the traces of the test from the DOM
+  owner, // the owner running the hook to use with `runWithOwner()` 
+} = renderResult
+```
+
+A primitive that manages the state of a counter could be tested like this:
+
+```ts frame="none"
+import { test, expect } from "vitest"
+import { renderHook } from "@solidjs/testing-library"
+import { createCounter } from "./counter"
+
+test("increments count", () => {
+  const { result } = renderHook(createCounter)
+  expect(result.count).toBe(0)
+  result.increment()
+  expect(result.count).toBe(1)
+})
+```
+
+
+## Testing effects
+
+Since effects may happen asynchronous, it can be difficult to test them. `@solidjs/testing-library` comes with a `testEffect` function that takes another function, which will receive a `done` function that can be called once the tests are over. It returns a promise that is resolved once done is called. Errors that would hit the next boundary are used to reject the returned promise.
+
+A test using it might look similar to this:
+
+```ts frame="none"
+const [value, setValue] = createSignal(0)
+return testEffect(done =>
+  createEffect((run: number = 0) => {
+    if (run === 0) {
+      expect(value()).toBe(0)
+      setValue(1)
+    } else if (run === 1) {
+      expect(value()).toBe(1)
+      done()
+    }
+    return run + 1
+  })
+)
+```
+
+## Integration/E2E testing
+
+Some issues can only be found once the code is running in the environment it is supposed to run in. Since integration and end-to-end tests are agnostic to frameworks, all proven approaches will work equally for Solid.
```

Comment:
```suggestion
Some issues can only be found once the code is running in the environment it is supposed to run in. 
Since integration and end-to-end tests are agnostic to frameworks, all proven approaches will work equally for Solid.
```